### PR TITLE
Categorized FAQs

### PR DESCRIPTION
**Pull request by singingwolfboy**
#### What are the relevant tickets?
#1190

#### What's this PR do?
Allow FAQ questions to be categorized

#### Where should the reviewer start?
Go into the CMS, click on the "Snippets" section, and make a FAQ category or two. Then add a few FAQs to a course, and assign them to categories

#### How should this be manually tested?
View the FAQ tab on the program page. Be sure to test with several different variations of questions and categories. Note that @roberthouse54 has asked me to just put the category in a `<h2>` tag and leave the styling to him, so the styling isn't very nice right now -- he will take care of that later (or perhaps push a commit to this pull request).

#### Screenshots (if appropriate)
snippets section
snippet FAQ categories
added a General FAQ category
selecting a category for a FAQ
setting up FAQs in a program
viewing FAQs in a program

### CODE REVIEW

**Comment by Ferdi** [comment]
@singingwolfboy thanks for the fast response on this.  This looks pretty good as a first pass.  
Is there anyway to (custom) order categories for program owners ?

**Comment by Ferdi** [comment]
Can we make "Category" a required field when present ?   They have the option to category="other"/"uncategorized".

**Comment by singingwolfboy** [comment]
@Ferdi: At the moment, there's no way to implement custom category orderings. If we support that, FAQ categories will have to be scoped to programs, so that different programs can specify different orderings for their FAQ categories. I'm not sure the best way to do that in Wagtail -- make it another page, instead of a snippet?

I don't really like the idea of making "Category" a required field. The way I've coded the pull request, if the program owner makes a question without specifying a category, it will end up in an "Uncategorized" section at the bottom. (See the screenshots I've posted.) It means that "Uncategorized" will always be the last category, even if the program owner makes categories with names that begin with "V", "X", "Y", or "Z". Does that do what you want?

**Comment by singingwolfboy** [comment]
Making "Category" an optional field also allows program owners to make uncategorized FAQs, which might be handy if the FAQ is very short. If there's only one or two questions, why bother to organize those questions into categories at all? If none of the questions have categories assigned to them, the template won't render any category `<h2>` tags, not even an "Uncategorized" one.

**Comment by pdpinch** [comment]
Should this be reviewed, or are we waiting for the changes described in https://github.com/mitodl/micromasters/issues/1190#issuecomment-251549015?

**Comment by Ferdi** [comment]
No need for review at the moment. DB will open a new PR

**Comment by singingwolfboy** [comment]
@Ferdi: Unfortunately, the changes we discussed may not be possible after all. I can easily set up the relationships between models so that each Program has many Categories and many FrequentlyAskedQuestions, and each FrequentlyAskedQuestion is associated with one Category, but the tricky part is making sure that questions are only associated with categories in the same program. Wagtail renders the Category dropdown menu with a list of _all_ categories across _all_ programs, which is very confusing: for every category in that list, there's no way to know which program it's associated with.

I pored through the documentation for both Django and Wagtail, and there doesn't seem to be any way to limit the association from the Django Models side. I asked several people about this online, and everyone I talked to said that the only way to do this is by customizing the Form class used to render the dropdown menu. However, Wagtail creates the Form class automatically, based on the Model! There is [a short section in Wagtail's documentation about customizing the generated forms](http://docs.wagtail.io/en/v1.6.3/advanced_topics/customisation/page_editing_interface.html#customising-generated-forms), but when I tried the code that the documentation suggested, it doesn't seem to actually *work*. I suspect there's some edge-case around customizing an _inline_ form, rather than a form for a full-page model. I went as far as reading form generation code within Wagtail, and it seems to account for inline forms, but the code is very complex, and I'm not familiar with it. There may well be a bug in how generated inline forms are customized.

I'm going to keep trying to get this to work, but I wanted to provide a status update, since I've spent several hours on this already. It may not be worth the time and effort.

**Comment by pdpinch** [comment]
Can we just make this really stupid and add a third field to the FAQ object: i.e., question, answer and category? We could leave it up to the view/template to handle the logic for grouping FAQs with the same category text.

**Comment by singingwolfboy** [comment]
@pdpinch: That's what's in this pull request right now. In this pull request, categories are global across all programs, and are always ordered alphabetically in the template. @Ferdi wanted me to make a new pull request where each program can define its own categories, and its own ordering for those categories. I thought that was possible, but it seems like I'm running up against the limitations of the Django ORM.

**Comment by singingwolfboy** [comment]
@pdpinch: oh, you're suggesting that we make "Category" just another text field on the FAQ object, rather than its own model?

**Comment by pdpinch** [comment]
Indeed, that's what I was suggesting. Maybe "denormalized" would have been clearer. 

I think @giocalitri has some thoughts. He said he'd get in touch after lunch -- in an hour or so?

**Comment by Ferdi** [comment]
Thanks @singingwolfboy 

Is it possible to override the form element generated by wagtail to only show categories that belong in the current program ? 

http://docs.wagtail.io/en/v1.6.3/advanced_topics/customisation/page_editing_interface.html#customising-generated-forms

**Comment by singingwolfboy** [comment]
@Ferdi see my comment above:

> There is [a short section in Wagtail's documentation about customizing the generated forms](http://docs.wagtail.io/en/v1.6.3/advanced_topics/customisation/page_editing_interface.html#customising-generated-forms), but when I tried the code that the documentation suggested, it doesn't seem to actually *work*. I suspect there's some edge-case around customizing an _inline_ form, rather than a form for a full-page model. I went as far as reading form generation code within Wagtail, and it seems to account for inline forms, but the code is very complex, and I'm not familiar with it. There may well be a bug in how generated inline forms are customized.

It may be possible, but it wasn't working when I followed the instructions in the documentation.

**Comment by singingwolfboy** [comment]
@pdpinch I made a new pull request with "category" as a simple text field: #1243

**Comment by singingwolfboy** [comment]
@Ferdi: if you want to see the code that I tried, I pushed the commit to GitHub: https://github.com/mitodl/micromasters/commit/14cce0725534046cbea9e690b25002b68693a5b7 As I said, it doesn't seem to work. It's also not clear how to customize the form to make the queryset dynamic, depending on which program is being rendered.

**Comment by pdpinch** [comment]
@singingwolfboy can you rebase this? (not sure how that happened...)

We'll move forward with this approach, but please open an issue for limiting the scope of the categories per program. Maybe you can try Wagtail support channels to see why your filter code wasn't working.

**Comment by pdpinch** [comment]
@amir-qayyum-khan can you review?

**Comment by singingwolfboy** [comment]
I've rebased this branch, and created #1266 as a follow-up issue.

**Comment by amir-qayyum-khan** [comment]
please rebase PR

**Comment by amir-qayyum-khan** [comment]
Functionality works as mentioned

File: cms/migrations/0019_add_faqcategory.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+from django.db import migrations, models
+import django.db.models.deletion
+import modelcluster.fields
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('cms', '0018_update_programfaculty'),
```
Comment (amir-qayyum-khan) [inline cms/migrations/0019_add_faqcategory.py]: Need to update this after rebase

Comment (amir-qayyum-khan) [inline cms/migrations/0019_add_faqcategory.py]: These is 0019 migration that i recently merged

**Comment by amir-qayyum-khan** [comment]
Code looks good

**Comment by pdpinch** [comment]
Now that #1259 has merged, we can approach this a little differently. If we add a category field to the FAQPage, we can write a template that combines all the FAQ pages into one, using the category from each one.

**Comment by pdpinch** [comment]
If #1332 works as expected, we'll replace this PR with that one.

**Pull request closed without merging**